### PR TITLE
Add Firewall Policy

### LINKED ISSUE #336: Feature Request - Add AZ Support for Azure Firewall in Secure vHub Model
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

**Is your feature request related to a problem?**

No

**Describe the solution you'd like**

AZs are now supported for Azure Firewalls deployed in the Secure vHub model (Virtual WAN) as per https://azure.microsoft.com/en-us/blog/boost-your-network-security-with-new-updates-to-azure-firewall & https://docs.microsoft.com/en-us/azure/virtual-wan/virtual-wan-faq#how-are-availability-zones-and-resiliency-handled-in-virtual-wan

This change looks to be to add the zones property to the Azure Firewall in the secure VWAN hub and expose as a variable and default to a sensible value (TBD). This is already done as part of the hub & spoke azure firewall so should be a simple addition before VWAN release.

**Additional context**

### PR DESCRIPTION
**Pull request by krowlandson**
## Overview/Summary

This PR adds an `azurerm_firewall_policy` resource for each `azurerm_firewall` created by the module.

To enable DNS proxy settings, each Azure Firewall needs to be associated with an Azure Firewall Policy. The Azure Firewall Policy also enables other firewall settings to be configured, including:

- `base_policy_id`
- `private_ip_ranges`
- `sku` (_should match the firewall sku_)
- `threat_intelligence_mode`
- `threat_intelligence_allowlist`
- `identity`
- `insights`
- `intrusion_detection`

> Note: not all of these settings are exposed directly through the module

## This PR fixes/adds/changes/removes

1. Adds an `azurerm_firewall_policy` resource for each `azurerm_firewall` created by the module
2. Enables the `enable_dns_proxy` feature flag for Azure Firewall settings (enabled via the attached Firewall Policy), including the ability to set custom `dns_servers`
3. Updates the `debug` output for the `connectivity` child module to ensure all `locals` are included

> Although no issue is logged for adding the firewall policy, this requirement was identified as part of the work for #287

4. Add the ability to set names for the `azurerm_public_ip` resource types (previously missing) as discussed in #332
5. Fix #336

### Breaking Changes

1. Added the following new settings to the `configure_connectivity_resources.settings.hub_networks[*].config.azure_firewall.config` input object to enable configuration of common settings for the new `azurerm_firewall_policy` resources:
    1. `dns_servers`
    2. `base_policy_id`
    3. `private_ip_ranges`
    4. `threat_intelligence_mode`
    5. `threat_intelligence_allowlist`
6. As part of addressing #332, we also fixed some addressing inconsistencies where the `name` field was being addressed by a different `scope` than the other properties. To improve consistency and flatten the `scope` object, we have updated the following resources:
    1. Setting the `name` field for `azurerm_virtual_network_gateway["expressroute"]` has been updated to `azurerm_virtual_network_gateway["connectivity_expressroute"]`
    7. Setting all other fields for `azurerm_virtual_network_gateway["connectivity"]["ergw"]` have been updated to `azurerm_virtual_network_gateway["connectivity_expressroute"]`
    8. Setting all fields for `azurerm_public_ip["connectivity"]["ergw"]` have been updated to `azurerm_public_ip["connectivity_expressroute"]`
    9. Setting the `name` field for `azurerm_virtual_network_gateway["vpngw"]` has been updated to `azurerm_virtual_network_gateway["connectivity_vpn"]`
    10. Setting all other fields for `azurerm_virtual_network_gateway["connectivity"]["vpngw"]` have been updated to `azurerm_virtual_network_gateway["connectivity_vpn"]`
    11. Setting all fields for `azurerm_public_ip["connectivity"]["vpn"]` have been updated to `azurerm_public_ip["connectivity_vpn"]`
    12. Setting all fields for `azurerm_public_ip["connectivity"]["azfw"]` have been updated to `azurerm_public_ip["connectivity_firewall"]`

> **IMPORTANT:** Customers who have already used the `advanced` settings to set custom values on any of the above resource types must take care when upgrading. Please review your code and verify your `plan` carefully when upgrading to avoid unexpected reconfiguration or redeployment of resources.

The following is an example showing the required `config` object for `azure_firewall` settings:

```hcl
  azure_firewall = {
    enabled = true
    config = {
      address_prefix                = "10.100.0.0/24"
      enable_dns_proxy              = true             # NOW WORKING
      dns_servers                   = []               # NEW
      sku_tier                      = ""               # NEW
      base_policy_id                = ""               # NEW
      private_ip_ranges             = []               # NEW
      threat_intelligence_mode      = ""               # NEW
      threat_intelligence_allowlist = []               # NEW
      availability_zones = {
        zone_1 = true
        zone_2 = true
        zone_3 = true
      }
    }
  }
```

The following is an example advanced config object reflecting some of these changes:

```hcl
    advanced = {
      custom_settings_by_resource_type = {
        azurerm_public_ip = {
          connectivity_expressroute = {
            (var.location) = {
              name = "pip-ergw-${var.location}"
            }
          }
          connectivity_vpn = {
            (var.location) = {
              name = "pip-vpngw-${var.location}"
            }

          }
          connectivity_firewall = {
            (var.location) = {
              name = "pip-fw-${var.location}"
            }
          }
        }
        azurerm_firewall = {
          connectivity = {
            (var.location) = {
              name = "fw-${var.location}"
            }
          }
          virtual_wan = {
            (var.location) = {
              name = "fw-hub-${var.location}"
            }
          }
        }
        azurerm_firewall_policy = {
          connectivity = {
            (var.location) = {
              name = "fwp-${var.location}"
            }
          }
          virtual_wan = {
            (var.location) = {
              name = "fwp-hub-${var.location}"
            }
          }
        }
      }
    }
```

## Testing Evidence

Please provide any testing evidence to show that your Pull Request works/fixes as described and planned (include screenshots, if appropriate).

azurerm_firewall_policy_outputs

azurerm_firewall_policy_for_hub

azurerm_firewall_policy_for_vwan

Evidence of DNS changes

Evidence of name changes

## As part of this Pull Request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/pulls)
- [x] Associated it with relevant [issues](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/issues), for tracking and closure.
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/tree/main)
- [x] Performed testing and provided evidence.
- [x] Updated relevant and associated documentation.
- [ ] Updated the ["What's New?"](https://github.com/Azure/Enterprise-Scale/wiki/Whats-new) wiki page (located in the [Enterprise-Scale repo](https://github.com/Azure/Enterprise-Scale) in the directory: `/docs/wiki/whats-new.md`)

### CODE REVIEW

File: modules/connectivity/locals.tf [hunk 1]
```diff
@@ -178,6 +178,13 @@ locals {
     local.deploy_virtual_network_gateway[location] &&
     hub_network.config.virtual_network_gateway.config.gateway_sku_vpn != local.empty_string
   }
+  deploy_azure_firewall_policy = {
+    for location, hub_network in local.hub_networks_by_location :
+    location =>
+    local.deploy_hub_network[location] &&
+    hub_network.config.azure_firewall.enabled &&
+    length(try(local.custom_settings.azurerm_firewall["connectivity"][location].firewall_policy_id, "")) == 0
```
Comment (matt-FFFFFF) [inline modules/connectivity/locals.tf:186]: Does this length comparison work when the try fails? E.g. if the `firewall_policy_id` is not set

File: modules/connectivity/locals.tf [hunk 2]
```diff
@@ -222,6 +229,13 @@ locals {
     local.deploy_virtual_hub[location] &&
     virtual_hub.config.vpn_gateway.enabled
   }
+  deploy_virtual_hub_azure_firewall_policy = {
+    for location, virtual_hub in local.virtual_hubs_by_location :
+    location =>
+    local.deploy_virtual_hub[location] &&
+    virtual_hub.config.azure_firewall.enabled &&
+    length(try(local.custom_settings.azurerm_firewall["virtual_wan"][location].firewall_policy_id, "")) == 0
```
Comment (matt-FFFFFF) [inline modules/connectivity/locals.tf:237]: Again, if the try fails does the length comparison work?

File: modules/connectivity/locals.tf (hunk 1 above)
Comment (krowlandson) [reply modules/connectivity/locals.tf:186]: In short, yes as the `try()` is there to determine when `firewall_policy_id` isn't set, and then returns `""` which evaluates to length = `0`

File: modules/connectivity/locals.tf (hunk 2 above)
Comment (krowlandson) [reply modules/connectivity/locals.tf:237]: as above

**Review by matt-FFFFFF — APPROVED** [review]
:shipit:

**Pull request merged**